src/rrdp.rs

```diff
@@ -1,456 +1,962 @@
-    processor.process(reader).map_err(|err| {
-        anyhow!("Error parsing snapshot XML: {:?}", &err)
-    })?;
-    Ok(processor)
-}
+    /// Cleans deprecated files and their parent directories if they are empty
+    pub fn clean(&mut self, config: &Config) -> Result<()> {
+        let clean_before = Time::seconds_ago(config.cleanup_after);
+
+        for deprecated in self
+            .deprecated_files
+            .iter()
+            .filter(|d| d.since < clean_before)
+        {
+            let path = &deprecated.path;
+            if path.exists() {
+                info!(
+                    "Removing RRDP file: {:?}, deprecated since: {}",
+                    path, deprecated.since
+                );
+                file_ops::remove_file_and_empty_parent_dirs(path)?;
+            }
+        }
 
-// DeltaParser parses RRDP delta XML and adds/removes records to/from the
-// referenced snapshot for publish/withdraw delta records respectively. As the
-// DeltaParser doesn't own the snapshot it needs a mutable reference to it which
-// is guaranteed (by 'a) to be valid as long as the DeltaParser instance exists.
-struct DeltaParser<'a> {
-    pub meta: Option<ParsedMeta>,
-    pub snapshot: &'a mut SnapshotParser,
-}
+        self.deprecated_files.retain(|d| d.since > clean_before);
 
-#[derive(Debug)]
-enum DeltaParserError {
-    Xml(XmlDecodeError),
-    MismatchedSessionId(Uuid),
-    InternalErrorMissingDeltaMeta,
-    InternalErrorMissingSnapshotMeta,
-    ItemToWithdrawNotFound(uri::Rsync),
-}
+        Ok(())
+    }
+
+    /// Persist the RRDP state to disk (as json)
+    pub fn persist(&self, path: &Path) -> Result<()> {
+        let json = serde_json::to_string_pretty(&self)?;
 
-impl From<XmlDecodeError> for DeltaParserError {
-    fn from(err: XmlDecodeError) -> Self {
-        DeltaParserError::Xml(err)
+        file_ops::write_buf(path, json.as_bytes()).with_context(|| "Could not save state.")
     }
-}
 
-impl<'a> DeltaParser<'a> {
-    fn new(snapshot: &'a mut SnapshotParser) -> Self {
-        DeltaParser {
-            meta: None,
-            snapshot,
-        }
+    pub fn session_id(&self) -> Uuid {
+        self.session_id
     }
 
-    fn is_applicable(&self) -> Result<bool, DeltaParserError> {
-        let self_meta = self.meta.as_ref().ok_or(
-            DeltaParserError::InternalErrorMissingDeltaMeta)?;
+    pub fn serial(&self) -> u64 {
+        self.serial
+    }
+
+    pub fn elements(&self) -> impl Iterator<Item = &CurrentObject> {
+        self.current_objects.objects()
+    }
 
-        let snapshot_meta = self.snapshot.meta.as_ref().ok_or(
-            DeltaParserError::InternalErrorMissingSnapshotMeta)?;
+    /// Writes the notification file to disk. Will first write to a
+    /// temporary file and then rename it to avoid serving partially
+    /// written files.
+    pub fn write_notification(&self) -> Result<()> {
+        let tmp_path = self.rrdp_dir.join(".notification.xml");
```

Comment:
Typically "." prefixes signify a hidden file but this is a temporary file, should it perhaps use a ".tmp" suffix instead?

Reply:
ack